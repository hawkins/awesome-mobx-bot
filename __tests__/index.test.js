// @ts-nocheck
const nock = require("nock");
const fs = require("fs");
const path = require("path");

const { Probot } = require("probot");

const app = require("../index.js");
const newLinkPayload = require("./fixtures/issue_new_link.json");
const existingLinkPayload = require("./fixtures/issue_existing_link.json");
const tools = require("../src/tools");

const awesomeMobXSource = `# Comparisons with other state management libraries

* [link](https://example.com/) - cool link 1!
* [link](https://example.com/) - cool link 2!
* [link](https://example.com/) - cool link 3!
`;
process.env.AWESOME_MOBX_SOURCE = awesomeMobXSource;

describe("issues", () => {
  let probot;
  let github;
  let mockCert;

  beforeAll(done => {
    fs.readFile(path.join(__dirname, "fixtures/mock-cert.pem"), (err, cert) => {
      if (err) return done(err);
      mockCert = cert;
      done();
    });
  });

  beforeEach(() => {
    nock.disableNetConnect();
    probot = new Probot({ id: 123, cert: mockCert });
    probot.load(app);

    github = {
      issues: { createComment: jest.fn(), update: jest.fn() },
      pullRequests: {
        create: jest.fn().mockReturnValue({
          data: { html_url: "https://example.com/1", number: 4 }
        })
      },
      gitdata: {
        getReference: jest
          .fn()
          .mockReturnValue({ data: { object: { sha: "01" } } }),
        createBlob: jest.fn().mockReturnValue({
          data: { sha: "a2", url: "https://example.com/2" }
        }),
        getTree: jest.fn().mockReturnValue({ data: { sha: "b1" } }),
        createTree: jest.fn().mockReturnValue({ data: { sha: "b2" } }),
        createCommit: jest.fn().mockReturnValue({ data: { sha: "c2" } }),
        createReference: jest
          .fn()
          .mockReturnValue({ data: { ref: "heads/bot/issue-3" } })
      }
    };

    // Passes the mocked out GitHub API into out robot instance
    probot.auth = () => Promise.resolve(github);
  });

  it("when opened with a new link", async () => {
    await probot.receive(newLinkPayload);
    expect(github.issues.createComment.mock.calls).toMatchSnapshot();
    expect(github.gitdata.createBlob.mock.calls).toMatchSnapshot();
    expect(github.gitdata.createCommit.mock.calls).toMatchSnapshot();
    expect(github.pullRequests.create.mock.calls).toMatchSnapshot();
  });

  it("when opened with an existing link", async () => {
    await probot.receive(existingLinkPayload);
    expect(github.issues.createComment.mock.calls).toMatchSnapshot();
    expect(github.issues.edit.mock.calls).toMatchSnapshot();
  });
});

describe("tools", () => {
  it("getSourceIndex can find the index of a type section", () => {
    expect(
      tools.getSourceIndex(
        awesomeMobXSource,
        "Comparisons with other state management libraries"
      )
    ).toMatchSnapshot();
    expect(
      tools.getSourceIndex(awesomeMobXSource, "unknown")
    ).toMatchSnapshot();
  });
});

/*
describe('My Probot app', () => {
  let probot
  let mockCert

  beforeAll((done) => {
    fs.readFile(path.join(__dirname, 'fixtures/mock-cert.pem'), (err, cert) => {
      if (err) return done(err)
      mockCert = cert
      done()
    })
  })

  beforeEach(() => {
    nock.disableNetConnect()
    probot = new Probot({ id: 123, cert: mockCert })
    // Load our app into probot
    probot.load(myProbotApp)
  })

  test('creates a comment when an issue is opened', async () => {
    // Test that we correctly return a test token
    nock('https://api.github.com')
      .post('/app/installations/2/access_tokens')
      .reply(200, { token: 'test' })

    // Test that a comment is posted
    nock('https://api.github.com')
      .post('/repos/hiimbex/testing-things/issues/1/comments', (body) => {
        expect(body).toMatchObject(issueCreatedBody)
        return true
      })
      .reply(200)

    // Receive a webhook event
    await probot.receive({ name: 'issues', payload })
  })

  afterEach(() => {
    nock.cleanAll()
    nock.enableNetConnect()
  })
})

// For more information about testing with Jest see:
// https://facebook.github.io/jest/

// For more information about testing with Nock see:
// https://github.com/nock/nock
*/
