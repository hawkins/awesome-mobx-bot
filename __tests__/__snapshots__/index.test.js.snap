// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issues when opened with a new link 1`] = `
Array [
  Array [
    Object {
      "body": "Hi there!

Thanks for suggesting that link! We found the following details for it:

- [MobX vs Redux with React: A noob’s comparison and questions](https://codeburst.io/mobx-vs-redux-with-react-a-noobs-comparison-and-questions-382ba340be09)
- Category: Comparisons with other state management libraries

I also opened an automatic PR with this link added. You can check this out [at PR #4](https://example.com/1) and make sure it looks good! :smile:

---

> Hi there, I'm a [bot](https://github.com/hawkins/awesome-mobx-bot) here to help automate tasks.

> If you have any questions, comments, or crude remarks for me, please leave them in [my project's repo](https://github.com/hawkins/awesome-mobx-bot/issues/new)

",
      "number": 3,
      "owner": "hawkins",
      "repo": "awesome-mobx-bot",
    },
  ],
]
`;

exports[`issues when opened with a new link 2`] = `
Array [
  Array [
    Object {
      "content": "# Comparisons with other state management libraries

* [link](https://example.com/) - cool link 1!
* [link](https://example.com/) - cool link 2!
* [link](https://example.com/) - cool link 3!
* [MobX vs Redux with React: A noob’s comparison and questions](https://codeburst.io/mobx-vs-redux-with-react-a-noobs-comparison-and-questions-382ba340be09)
",
      "encoding": "utf-8",
      "owner": "hawkins",
      "repo": "awesome-mobx-bot",
    },
  ],
]
`;

exports[`issues when opened with a new link 3`] = `
Array [
  Array [
    Object {
      "message": "Add 'MobX vs Redux with React: A noob’s comparison and questions' by hawkins

Fixes issue #3
",
      "owner": "hawkins",
      "parents": Array [
        "b1",
      ],
      "repo": "awesome-mobx-bot",
      "tree": "b2",
    },
  ],
]
`;

exports[`issues when opened with a new link 4`] = `
Array [
  Array [
    Object {
      "base": "master",
      "body": "Fixes issue #3

This is an automatic pull request spawned by the above issue.

---

> Hi there, I'm a [bot](https://github.com/hawkins/awesome-mobx-bot) here to help automate tasks.

> If you have any questions, comments, or crude remarks for me, please leave them in [my project's repo](https://github.com/hawkins/awesome-mobx-bot/issues/new)
",
      "head": "bot/issue-3",
      "owner": "hawkins",
      "repo": "awesome-mobx-bot",
      "title": "Add 'MobX vs Redux with React: A noob’s comparison and questions' by hawkins",
    },
  ],
]
`;

exports[`issues when opened with an existing link 1`] = `
Array [
  Array [
    Object {
      "body": "Hey, thanks so much for suggesting that link! :+1:

Turns out we've already got that link in the list, so I'm closing this issue since it's a duplicate. Thanks so much though! :raised_hands:

---

> Hi there, I'm a [bot](https://github.com/hawkins/awesome-mobx-bot) here to help automate tasks.

> If you have any questions, comments, or crude remarks for me, please leave them in [my project's repo](https://github.com/hawkins/awesome-mobx-bot/issues/new)

",
      "number": 3,
      "owner": "hawkins",
      "repo": "awesome-mobx-bot",
    },
  ],
]
`;

exports[`issues when opened with an existing link 2`] = `
Array [
  Array [
    Object {
      "number": 3,
      "owner": "hawkins",
      "repo": "awesome-mobx-bot",
      "state": "closed",
    },
  ],
]
`;

exports[`tools getSourceIndex can find the index of a type section 1`] = `
Object {
  "index": 0,
  "length": 190,
}
`;

exports[`tools getSourceIndex can find the index of a type section 2`] = `
Object {
  "index": -1,
  "length": -1,
}
`;
